test/common/conn_pool/conn_pool_base_test.cc

```diff
@@ -202,6 +242,188 @@ TEST_F(ConnPoolImplBaseTest, ExplicitPreconnectNotHealthy) {
+  // this pool manually.
+  pool_.drainConnectionsImpl(Envoy::ConnectionPool::DrainBehavior::DrainAndDelete);
+}
+
+TEST_F(ConnPoolImplDispatcherBaseTest, MaxConnectionDurationTimerEnabled) {
+  ON_CALL(*cluster_, maxConnectionDuration)
+      .WillByDefault(Return(absl::optional<std::chrono::milliseconds>(max_connection_duration_)));
+
+  // Create a new stream using the pool
+  EXPECT_CALL(pool_, instantiateActiveClient);
+  pool_.newStreamImpl(context_);
+  ASSERT_EQ(1, clients_.size());
+
+  // Expect the lifetime timer to only be enabled after connecting.
+  EXPECT_CALL(pool_, onPoolReady);
+  EXPECT_EQ(ActiveClient::State::CONNECTING, clients_.back()->state());
+  EXPECT_EQ(nullptr, clients_.back()->lifetime_timer_);
+
+  clients_.back()->onEvent(Network::ConnectionEvent::Connected);
+  EXPECT_EQ(ActiveClient::State::BUSY, clients_.back()->state());
+  EXPECT_EQ(true, clients_.back()->lifetime_timer_->enabled());
+
+  // Close active stream and expect that the client goes back to ready
+  clients_.back()->active_streams_ = 0;
+  pool_.onStreamClosed(*clients_.back(), false);
+  EXPECT_EQ(ActiveClient::State::READY, clients_.back()->state());
+
+  // We are only testing that the timer was enabled. So, to clean up, just disable
+  // the timer and drain this pool manually.
+  clients_.back()->lifetime_timer_->disableTimer();
+  pool_.drainConnectionsImpl(Envoy::ConnectionPool::DrainBehavior::DrainAndDelete);
+}
+
+TEST_F(ConnPoolImplDispatcherBaseTest, MaxConnectionDurationBusy) {
+  ON_CALL(*cluster_, maxConnectionDuration)
+      .WillByDefault(Return(absl::optional<std::chrono::milliseconds>(max_connection_duration_)));
+
+  // Create a new stream using the pool
+  EXPECT_CALL(pool_, instantiateActiveClient);
+  pool_.newStreamImpl(context_);
+  ASSERT_EQ(1, clients_.size());
+
+  // Connect and expect busy (1 stream per connection)
+  EXPECT_CALL(pool_, onPoolReady);
+  EXPECT_EQ(ActiveClient::State::CONNECTING, clients_.back()->state());
+  clients_.back()->onEvent(Network::ConnectionEvent::Connected);
+  EXPECT_EQ(ActiveClient::State::BUSY, clients_.back()->state());
+
+  // Verify that advancing to just before the lifetime timeout doesn't drain the connection.
+  time_system_.advanceTimeAndRun(std::chrono::milliseconds(max_connection_duration_ - 1),
+                                 *dispatcher_, Event::Dispatcher::RunType::Block);
+  EXPECT_EQ(0, pool_.host()->cluster().stats().upstream_cx_max_duration_.value());
+  EXPECT_EQ(ActiveClient::State::BUSY, clients_.back()->state());
+
+  // Verify that advancing past the lifetime timeout drains the connection,
+  // because there's a busy client.
+  time_system_.advanceTimeAndRun(std::chrono::milliseconds(2), *dispatcher_,
+                                 Event::Dispatcher::RunType::Block);
+  EXPECT_EQ(1, pool_.host()->cluster().stats().upstream_cx_max_duration_.value());
+  EXPECT_EQ(ActiveClient::State::DRAINING, clients_.back()->state());
```

Comment:
check that the stat you added gets bumped?

Reply:
That check is at L327 unless you had another stat in mind